analyses/notebooks/rl_intro_notebook.ipynb

```diff
@@ -0,0 +1,2171 @@
+    "\n",
+    "    def get_best_action(self, state: np.ndarray) -> np.ndarray:\n",
+    "        \"\"\"\n",
+    "        Gets the action which the agent thinks is best in the current state\n",
+    "\n",
+    "        Args:\n",
+    "            state: the current state the agent is in\n",
+    "\n",
+    "        Returns:\n",
+    "            the action with the highest q_value\n",
+    "        \"\"\"\n",
+    "        # get the action corresponding to the highest q_value given the state\n",
+    "        return max(self.table[state], key=self.table[state].get)\n",
+    "\n",
+    "    def print_q_table(self) -> pd.DataFrame:\n",
+    "        \"\"\"\n",
+    "        Prints the Q-table in a readable format\n",
+    "        \n",
+    "        Returns:\n",
+    "            the q-table in data frame format\n",
+    "        \"\"\"\n",
+    "    \n",
+    "        # convert the dictionary q-table to a data frame\n",
+    "        df = pd.DataFrame.from_dict(self.table, orient='index')\n",
+    "\n",
+    "        # overwrite columns and indices by string for readability\n",
+    "        df.columns = [str(column) for column in df.columns]\n",
+    "        df.index = [str(index) for index in df.index]\n",
+    "\n",
+    "        # return the q-table as dataframe\n",
+    "        return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "TkK0fjvdFXwk"
+   },
+   "source": [
+    "## Agent\n",
+    "As can be seen in the figure above (in the Environment section), the agent receives the state and reward from the environment and returns an action. Hence, the two attributes of the agent class are the action it chooses and the Q_table class, which provides the intelligence of choosing an action. The only function this class has is ```choose_action(self, state, action_range, eval)```. Depending on whether the agent is evaluating, the agent chooses an action randomly from the action range (```eval = False```) or it chooses the action the agent thinks is best (```eval = True```). This function can be expanded to include more exploration-exploitation methods, e.g. Epsilon greedy or Boltzmann."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "id": "ivw3_rSJ-E1S"
+   },
+   "outputs": [],
+   "source": [
+    "class Agent():\n",
+    "    \"\"\"\n",
+    "    Class representing the agent\n",
+    "\n",
+    "    Arguments:\n",
+    "        action: the action performed by the agent\n",
+    "        q_table: instance of the Q_table class, which represents the intelligence of the agent\n",
+    "    \"\"\"\n",
+    "    def __init__(self, par: Parameters):\n",
+    "        self.action = np.zeros([par.action_size])\n",
+    "        self.q_table = Q_table(par)\n",
+    "        print('Agent class initialized...')\n",
+    "\n",
+    "    def choose_action(self, state: np.ndarray, action_range: np.ndarray, eval: bool) -> np.ndarray:\n",
```

Comment:
choose_action could fail if neither of the if-statements are valid. Is that the desired result? Should there be a closing else: raise ValueError?